src/session-lock.c

```diff
@@ -211,130 +210,125 @@ session_lock_output_create(struct session_lock *lock, struct output *output)
 }
 
 static void
-session_lock_destroy(struct session_lock *lock)
+session_lock_destroy(struct session_lock_manager *manager)
 {
 	struct session_lock_output *lock_output, *next;
-	wl_list_for_each_safe(lock_output, next, &lock->session_lock_outputs, link) {
+	wl_list_for_each_safe(lock_output, next, &manager->session_lock_outputs, link) {
 		wlr_scene_node_destroy(&lock_output->tree->node);
 	}
-	if (g_server->session_lock == lock) {
-		g_server->session_lock = NULL;
+	if (manager->lock) {
+		wl_list_remove(&manager->lock_destroy.link);
+		wl_list_remove(&manager->lock_unlock.link);
+		wl_list_remove(&manager->lock_new_surface.link);
 	}
-	if (!lock->abandoned) {
-		wl_list_remove(&lock->destroy.link);
-		wl_list_remove(&lock->unlock.link);
-		wl_list_remove(&lock->new_surface.link);
-	}
-	free(lock);
+	manager->lock = NULL;
```

Comment:
Should we not `free(manager->lock)`?

Reply:
`manager->lock` is just `wlr_session_lock_v1` so wlroots handles it.

Sway wraps `wlr_session_lock_v1` with `sway_session_lock` and separates it from session lock manager like:
```c
struct sway_session_lock {
	struct wlr_session_lock_v1 *lock;
	struct wlr_surface *focused;
	bool abandoned;

	struct wl_list outputs; // struct sway_session_lock_output

	// invalid if the session is abandoned
	struct wl_listener new_surface;
	struct wl_listener unlock;
	struct wl_listener destroy;
};
```
```c
struct sway_server {
	...
	struct {
		struct sway_session_lock *lock;
		struct wlr_session_lock_manager_v1 *manager;

		struct wl_listener new_lock;
		struct wl_listener manager_destroy;
	} session_lock;
	...
}
```

So sway frees `sway_session_lock` every time a session-lock is destroyed, but I prefer just creating `session_lock_manager` which is persistent throughout the session because there cannot be more than 1 session-lock:
```c
struct session_lock_manager {
	struct server *server;
	struct wlr_session_lock_manager_v1 *wlr_manager;
	struct wlr_session_lock_v1 *lock;
	struct wlr_surface *focused;
	bool locked;

	struct wl_list session_lock_outputs;

	struct wl_listener new_lock;
	struct wl_listener destroy;

	struct wl_listener lock_new_surface;
	struct wl_listener lock_unlock;
	struct wl_listener lock_destroy;
};
```

Reply:
Great table. I think we should add the second table as a comment in the source file. It's likely to be a useful reminder in years to come :smile:

Reply:
Added:
```c
	/*
	 * When not locked: lock=NULL, locked=false
	 * When locked: lock=non-NULL, locked=true
	 * When lock is destroyed without being unlocked: lock=NULL, locked=true
	 */
```